contrib/nydus-backend-proxy/src/main.rs

```diff
@@ -0,0 +1,240 @@
+            Outcome::Success(HeaderData {
+                _host: host,
+                range: rangestr,
+            })
+        }
+    }
+
+    #[rocket::head("/<_namespace>/<_repo>/blobs/<digest>")]
+    pub async fn blob_check(
+        _namespace: PathBuf,
+        _repo: PathBuf,
+        digest: String,
+    ) -> Result<Option<FileStream>, Status> {
+        let dis = &digest[7..];
+        let mut path = IMAGEPATH.clone();
+        path.push(&dis);
+        Ok(NamedFile::open(path)
+            .await
+            .ok()
+            .map(|nf| FileStream(nf, dis.to_string())))
+    }
+
+    /* fetch blob response
+     * NamedFile: blob data
+     * String: Docker-Content-Digest
+     */
+    pub struct FileStream(NamedFile, String);
+
+    impl<'r> Responder<'r, 'static> for FileStream {
+        fn respond_to(self, req: &'r Request<'_>) -> response::Result<'static> {
+            Response::build_from(self.0.respond_to(req)?)
+                .raw_header("Docker-Content-Digest", self.1)
+                .raw_header("Content-Type", "application/octet-stream")
+                .ok()
+        }
+    }
+
+    /* fetch blob part response(stream)
+     * path: path of blob
+     * dis: Docker-Content-Digest
+     * start & end: "Content-Range: bytes <start>-<end>/<size>"
+     */
+    pub struct RangeStream {
+        dis: String,
+        start: u64,
+        len: u64,
+    }
+
+    impl RangeStream {
+        fn get_rangestr(&self) -> String {
+            let endpos = self.start + self.len - 1;
+            format!("bytes {}-{}/{}", self.start, endpos, self.len)
+        }
+    }
+
+    impl<'r> Responder<'r, 'static> for RangeStream {
+        fn respond_to(self, _req: &'r Request<'_>) -> response::Result<'static> {
+            let file = match BLOBFDMAP.get(&ffi::OsString::from(&self.dis)) {
+                Some(fd) => fd,
+                None => return Err(Status::NotFound),
+            };
+            let rangestring = self.get_rangestr();
+            let size = self.len;
+            const BUFSIZE: usize = 4096;
+            let mut buf = vec![0; BUFSIZE];
+            let mut read = 0;
+            let startpos = self.start as i64;
+            let raw_fd = file.as_raw_fd();
+            Response::build()
+                .streamed_body(ReaderStream! {
+                    while read < size
+                    {
+                        let unread = (size - read) as usize;
+                        let allbuf: &mut [u8] = &mut buf[..];
+                        match uio::pread(raw_fd, allbuf, startpos+read as i64)
+                        {
+                            Ok(n) =>
+                            {
+                                if n == 0
+                                {
+                                    break
+                                }
+                                read += n as u64;
+                                if unread < BUFSIZE {
+                                let part = &allbuf[0..unread];
```

Comment:
Tidy codes with `cargo fmt`.

Reply:
Are there any changes here?

Reply:
> Are there any changes here?

I have tried to tidy codes with `cargo fmt`, but it seems nothing change here. It seems that Rustfmt does not work where macro declarations and uses. Maybe I should tidy the code here manually.